src/crop-manager.sol

```diff
@@ -0,0 +1,171 @@
+
+interface VatLike {
+    function live() external view returns (uint256);
+    function urns(bytes32, address) external view returns (uint256, uint256);
+    function dai(address) external view returns (uint256);
+    function fork(bytes32, address, address, int256, int256) external;
+    function frob(bytes32, address, address, address, int256, int256) external;
+    function flux(bytes32, address, address, uint256) external;
+    function hope(address) external;
+    function nope(address) external;
+}
+
+interface CropLike {
+    function gem() external view returns (address);
+    function ilk() external view returns (bytes32);
+    function join(address, address, uint256) external;
+    function exit(address, address, uint256) external;
+    function tack(address, address, uint256) external;
+    function flee(address) external;
+}
+
+interface TokenLike {
+    function approve(address, uint256) external;
+    function transferFrom(address, address, uint256) external;
+}
+
+contract UrnProxy {
+    address immutable public vat;
+    address immutable public usr;
+    address public owner;
+
+    constructor(address vat_, address usr_) public {
+        owner = msg.sender;
+        vat = vat_;
+        usr = usr_;
+        VatLike(vat_).hope(msg.sender);
+    }
+
+    function migrate(address newOwner) external {
+        require(msg.sender == owner, "UrnProxy/not-owner");
+        VatLike(vat).hope(newOwner);
+        VatLike(vat).nope(owner);
+        owner = newOwner;
+    }
+}
+
+contract CropJoinManager {
+    mapping (address => uint256) public wards;
+    mapping (address => address) public proxy;  // UrnProxy per user
+    address public implementation;
+
+    event Rely(address indexed);
+    event Deny(address indexed);
+    event SetImplementation(address indexed);
+
+    constructor() public {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function rely(address usr) public auth { wards[usr] = 1; emit Rely(msg.sender); }
+    function deny(address usr) public auth { wards[usr] = 0; emit Deny(msg.sender); }
+    modifier auth { require(wards[msg.sender] == 1, "CropJoinManager/non-authed"); _; }
+
+
+    function setImplementation(address implementation_) external auth {
+        implementation = implementation_;
+        emit SetImplementation(implementation_);
+    }
+
+    fallback() external {
+        address _impl = implementation;
+        require(_impl != address(0));
+
+        assembly {
+            let ptr := mload(0x40)
+            calldatacopy(ptr, 0, calldatasize())
+            let result := delegatecall(gas(), _impl, ptr, calldatasize(), 0, 0)
+            let size := returndatasize()
+            returndatacopy(ptr, 0, size)
+
+            switch result
+            case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+}
+
+contract CropJoinManagerImp {
+    bytes32 slot0;
+    mapping (address => address) public proxy;  // UrnProxy per user
```

Comment:
If this is already `public` on the proxy, I don't think it needs to be public here...it just adds to the bytecode size unnecessarily.

Reply:
It is convenient for the calling code. For example if I just set this to `internal` the test code will error: 

```
dapp --use solc:0.6.12 test --rpc -v --fuzz-runs 1
src/test/crop-manager-unit.sol:50:16: TypeError: Member "proxy" not found or not visible after argument-dependent lookup in contract CropJoinManagerImp.
        return manager.proxy(address(this));
               ^-----------^
```

I'm okay either way. Casting it to `CropJoinManager` is not a big deal.

Reply:
Conclusion: remove `public`, but keep mapping definition for convenience.

Reply:
Done.